keps/sig-node/3386-kubelet-evented-pleg/README.md

```diff
@@ -0,0 +1,546 @@
+
+### Non-Goals
+
+- Completely eliminate polling altogether.
+    - This proposal does not advocate completely removing the polling. We cannot solely rely on the upstream container events due to the possibility of missing events. PLEG should relist at reduced frequency to ensure no events are missed.
+- Addressing container image relisting via CRI events is out of scope for this enhancement at this point in time.
+
+## Proposal
+
+This proposal aims to replace the periodic polling with a pod lifecycle event watcher. Currently, the Kubelet calls into three CRI calls of the form `List*`: [ListContainers](https://github.com/kubernetes/kubernetes/blob/6efd6582df2011f1ec8c146ef711b3348ae07d60/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L78), [ListPodSandbox](https://github.com/kubernetes/kubernetes/blob/6efd6582df2011f1ec8c146ef711b3348ae07d60/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L60). Each of these is used to populate the Kubelet's perspective
+of the state of the node.
+
+As the number of pods on a node increases, the amount of time the Kubelet and CRI implementation takes in generating and reading this list increases linearly. What is needed is a way of the Kubelet being notified when a container changes state in a way it did not trigger.
+
+There should only be two such cases, and in normal operation, only one would happen frequently:
+
+- The first, and most clear case of a container changing state without the Kubelet triggering that state change is when a container stops. Containers can exit gracefully, or be OOM killed, and the Kubelet would not know.
+   - We will also introduce events when the container is created as well as is started. This will help us reduce the relisting that takes placed while the kubelet waits for the container to start.
+   - Although kubelet initiates the container deletion, for sake of increased validation we are also introducing the event to denote that from the runtime.
+- The second, and less likely case is when another entity comes and changes the state of the node.
+	- For container related events (such as a container creating, starting, stopping or being killed), this can appear as a user calling [crictl](https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/crictl.md) manually, or even using the runtime directly.
+
+The Kubelet currently covers each of thse cases quite easily: by listing all of the resources on the node, it will have an accurate picture after the amount of time of its [poll interval](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/kubelet.go#L162). For each of these cases, a new CRI-based events API can be made, using [gRPC server streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc). This way, the entity closest to the activity of the containers and pods (the CRI implementation) can be responsible for informing the Kubelet of their behavior directly.
```

Comment:
When a gRPC connection is closed (due to a kubelet failure), how does the Kubelet "catch up" to the latest events?

I would assume that we would have to "ListContainerStatuses", then start the events channel for all events that occur after the "List".  That requires coordination and buffering on the CRI side - a way to find all container statuses newer than the time the List was returned.  That ensures that in the event of a Kubelet failure we don't miss any events.

I expected to see it described here how that would be solved, but I'm missing it.  Is it because we're assuming the `created_at` attribute of the Container Event can be filtered on?  If so, how are we filtering that in the CRI and how is the Kubelet comparing the results of the List call to the start point of the events?  CRI `GetContainerEvents` does not have an argument that allows us to get "events after time T".

Can someone point me to where this is described?

Reply:
The current implementation in Generic PLEG uses [computeEvents](https://github.com/kubernetes/kubernetes/blob/497e0dfd33b180fcc3855aad794d87f0e429c82c/pkg/kubelet/pleg/generic.go#L351-L361) that is called upon periodic relisting and which generates PLEG events based upon last updated pod status present in cache.

Since, in case of Evented PLEG we aim to still re-use the relisting functionality of Generic PLEG (though with relisting being called at a higher frequency than usual 1s) and since the pod status(es) are persisted to common cache, that func should be still able to generate the events since the last update / gRPC connection failure [(which forces a relist as well)](https://github.com/kubernetes/kubernetes/pull/111384/files#diff-465aae123f6877816ff0428968fc74f503fe15b27b434480bf57b79e62c7fa2fR161).

Reply:
+1 to what Clayton suggests. 

A little bit more on  "how we expect to reliably ensure ordered deliver of events": We should be clear on how CRI should buffer container events. For example, if kubelet stops reading from grpc stream for any reason, we can't expect CRI to buffer the events indefinitely. Some events will be discarded if kubelet stops for too long. Kubelet should be able to handle this case.

Reply:
Thanks @ruiwen-zhao and @smarterclayton for these important points. We experimented with a fix that should address these issues, and will update KEP and the implementation as soon as possible.

Reply:
both kubelet and the container runtime may be restarted resulting in unexpected/unknown states of shim managed containers, esp. while the container runtime is "restarting" https://github.com/containerd/containerd/blob/main/pkg/cri/server/restart.go